src/librustdoc/clean/types.rs

```diff
@@ -1966,11 +1967,64 @@ crate struct Static {
 }
 
 #[derive(Clone, PartialEq, Eq, Hash, Debug)]
-crate struct Constant {
-    crate type_: Type,
-    crate expr: String,
-    crate value: Option<String>,
-    crate is_literal: bool,
+crate enum Constant {
+    /// Typed constant value.
+    TyConst { type_: Type, expr: String },
+    /// A constant (expression) that’s not an item or associated item. These are usually found
+    /// nested inside types (e.g., array lengths) or expressions (e.g., repeat counts), and also
+    /// used to define explicit discriminant values for enum variants.
+    Anonymous { type_: Type, body: BodyId },
+    /// Inlined constant.
+    Extern { type_: Type, did: DefId },
+    /// const FOO: u32 = ...;
+    Local { type_: Type, did: DefId, body: BodyId },
+}
+
+impl Constant {
+    crate fn expr(&self, tcx: TyCtxt<'_>) -> String {
+        match self {
+            Self::TyConst { expr, .. } => expr.clone(),
+            Self::Extern { did, .. } => print_inlined_const(tcx, *did),
+            Self::Local { body, .. } | Self::Anonymous { body, .. } => print_const_expr(tcx, *body),
+        }
+    }
+
+    crate fn value(&self, tcx: TyCtxt<'_>) -> Option<String> {
+        match self {
+            Self::TyConst { .. } | Self::Anonymous { .. } => None,
+            Self::Extern { did, .. } | Self::Local { did, .. } => print_evaluated_const(tcx, *did),
+        }
+    }
+
+    crate fn is_literal(&self, tcx: TyCtxt<'_>) -> bool {
+        match self {
+            Self::TyConst { .. } => false,
+            Self::Extern { did, .. } => did
+                .as_local()
+                .map_or(false, |did| is_literal_expr(tcx, tcx.hir().local_def_id_to_hir_id(did))),
+            Self::Local { body, .. } | Self::Anonymous { body, .. } => {
+                is_literal_expr(tcx, body.hir_id)
+            }
+        }
+    }
+
+    crate fn type_(&self) -> &Type {
+        match *self {
+            Self::TyConst { ref type_, .. }
+            | Self::Extern { ref type_, .. }
+            | Self::Local { ref type_, .. }
+            | Self::Anonymous { ref type_, .. } => type_,
+        }
+    }
+
+    crate fn to_type(self) -> Type {
+        match self {
+            Self::TyConst { type_, .. }
+            | Self::Extern { type_, .. }
+            | Self::Local { type_, .. }
+            | Self::Anonymous { type_, .. } => type_,
+        }
+    }
```

Comment:
Have you considered using a struct that holds the type instead? Then you wouldn't need these two functions and you could just access the type directly:
```rust
struct Constant {
  type_: Type,
  kind: ConstantKind,
}

enum ConstantKind { ... }
```

Reply:
I did but I wondered for a while if it was worth it or not. I personally think that the current code is fine, but if you prefer otherwise, I can update.

Reply:
I think I would rather have the explicit field, yeah.

Reply:
Then let's go for it!